File: backend/app/api/v1/endpoints/consent.py
"""
Vaidya — /api/v1/consent  (DPDP Act 2023 compliance)

POST /grant          — patient grants one or more consent types (explicit, informed)
POST /revoke         — patient revokes a previously granted consent
GET  /status/{patient_id} — current consent state for a patient
GET  /audit/{patient_id}  — full consent audit trail (immutable log)

Design rationale:
  DPDP Act 2023 §6 requires that consent be:
    - Free, specific, informed, unconditional, unambiguous
    - Obtained before any processing of personal data
    - Revocable at any time with equal ease
    - Linked to a specific purpose and policy version

  Three consent types map to three distinct processing purposes:
    data_processing      — core triage inference (required to use the app)
    anonymised_analytics — district-level outbreak detection (optional)
    asha_contact         — ASHA worker can contact patient for follow-up (optional)

  Each consent record is immutable — revocation writes a new record (granted=False)
  rather than deleting the original. This gives a full, auditable trail.

  The consent_version field ties each grant to the exact privacy notice the
  patient saw — if the notice changes, existing grants remain valid but users
  must re-consent for the new version.
"""

import hashlib
from datetime import datetime, timezone
from typing import List, Optional
from uuid import UUID

import structlog
from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user_id
from app.models.models import AuditLog, ConsentLog, Patient

router = APIRouter()
logger = structlog.get_logger(__name__)

# Current privacy notice version — bump this when the notice text changes
CURRENT_CONSENT_VERSION = "1.0"

# Human-readable descriptions shown to the patient before they grant consent
CONSENT_DESCRIPTIONS = {
    "data_processing": (
        "Vaidya will process your symptom description, voice recording, and/or "
        "medical image to provide an AI-assisted triage result. Your data is stored "
        "securely on servers in India and deleted after 90 days (raw text after 30 days). "
        "This consent is required to use the triage service."
    ),
    "anonymised_analytics": (
        "Your anonymised triage data (district, diagnosis category, triage level — "
        "no name, phone, or identifiers) will be included in district-level outbreak "
        "detection to help public health authorities respond faster to disease clusters. "
        "This is optional and does not affect your triage result."
    ),
    "asha_contact": (
        "Your assigned ASHA worker will receive your triage level and a follow-up "
        "reminder. They may contact you by phone to check on your recovery. "
        "This is optional. You can revoke it at any time."
    ),
}


# ── Schemas ────────────────────────────────────────────────────────────────────

class ConsentGrantRequest(BaseModel):
    patient_id: UUID
    consent_types: List[str] = Field(
        ...,
        min_length=1,
        description="List of consent types to grant",
    )
    policy_version: str = Field(
        default=CURRENT_CONSENT_VERSION,
        description="Privacy policy version the patient reviewed",
    )
    language: str = Field(
        default="en",
        pattern="^(en|hi|ta)$",
        description="Language in which the consent notice was displayed",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "patient_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
                "consent_types": ["data_processing", "asha_contact"],
                "policy_version": "1.0",
                "language": "ta",
            }
        }


class ConsentRevokeRequest(BaseModel):
    patient_id: UUID
    consent_type: str = Field(..., description="Consent type to revoke")
    reason: Optional[str] = Field(
        None,
        max_length=500,
        description="Optional reason for revocation (not stored, only logged)",
    )


class ConsentStatusItem(BaseModel):
    consent_type: str
    granted: bool
    version: str
    timestamp: datetime
    language: Optional[str] = None


class ConsentStatusResponse(BaseModel):
    patient_id: UUID
    consents: List[ConsentStatusItem]
    has_required_consent: bool      # data_processing must be granted to use triage
    evaluated_at: datetime


class ConsentAuditEntry(BaseModel):
    id: UUID
    consent_type: str
    granted: bool
    version: str
    timestamp: datetime
    ip_hash: Optional[str] = None   # SHA-256 of IP — never raw IP in response


class ConsentAuditResponse(BaseModel):
    patient_id: UUID
    entries: List[ConsentAuditEntry]
    total: int


# ── Helpers ────────────────────────────────────────────────────────────────────

def _validate_consent_types(types: List[str]) -> None:
    valid = set(CONSENT_DESCRIPTIONS.keys())
    invalid = [t for t in types if t not in valid]
    if invalid:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Unknown consent type(s): {invalid}. Valid: {sorted(valid)}",
        )


def _hash_ip(ip: Optional[str]) -> Optional[str]:
    """Hash IP address for storage — we log that consent was given from an IP,
    but never store the raw address to avoid creating linkable PII."""
    if not ip:
        return None
    return hashlib.sha256(ip.encode()).hexdigest()[:16]


async def _get_patient_or_404(patient_id: UUID, db: AsyncSession) -> Patient:
    result = await db.execute(select(Patient).where(Patient.id == patient_id))
    patient = result.scalar_one_or_none()
    if not patient:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Patient {patient_id} not found",
        )
    return patient


# ── POST /consent/grant ────────────────────────────────────────────────────────

@router.post(
    "/grant",
    status_code=status.HTTP_201_CREATED,
    summary="Grant consent for one or more processing purposes (DPDP §6)",
    response_description="List of consent records created",
)
async def grant_consent(
    payload: ConsentGrantRequest,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """
    Records explicit, informed consent from a patient before any data processing.

    Must be called with `data_processing` consent before the first triage session.

    Each call is idempotent for the same (patient, type, version) combination —
    re-granting the same consent at the same version is a no-op that returns the
    existing record rather than creating a duplicate.

    DPDP §6 compliance notes recorded:
    - Consent is specific (one row per purpose)
    - Policy version links grant to exact notice shown
    - IP address hashed (never raw)
    - Timestamp is UTC with timezone
    """
    _validate_consent_types(payload.consent_types)

    await _get_patient_or_404(payload.patient_id, db)

    ip_hash = _hash_ip(request.client.host if request.client else None)
    created = []

    for ct in payload.consent_types:
        # Idempotency check — skip if already granted at this version
        existing = await db.execute(
            select(ConsentLog).where(
                ConsentLog.patient_id == payload.patient_id,
                ConsentLog.consent_type == ct,
                ConsentLog.granted == True,      # noqa: E712
                ConsentLog.version == payload.policy_version,
            )
        )
        if existing.scalar_one_or_none():
            logger.debug(
                "vaidya.consent.already_granted",
                patient=str(payload.patient_id)[:8],
                type=ct,
            )
            continue

        record = ConsentLog(
            patient_id=payload.patient_id,
            consent_type=ct,
            granted=True,
            ip_address=ip_hash,
            version=payload.policy_version,
        )
        db.add(record)

        # Immutable audit trail
        audit = AuditLog(
            event_type="consent.granted",
            entity_type="consent_log",
            entity_id=str(payload.patient_id),
            actor_id=str(payload.patient_id),
            ip_address=ip_hash,
        )
        db.add(audit)
        created.append(ct)

    await db.commit()

    logger.info(
        "vaidya.consent.granted",
        patient=str(payload.patient_id)[:8],
        types=created,
        version=payload.policy_version,
    )

    return {
        "granted": created,
        "policy_version": payload.policy_version,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "message": (
            f"Consent recorded for: {', '.join(created)}"
            if created
            else "All specified consents were already granted at this policy version."
        ),
    }


# ── POST /consent/revoke ───────────────────────────────────────────────────────

@router.post(
    "/revoke",
    status_code=status.HTTP_200_OK,
    summary="Revoke a previously granted consent (DPDP §6 — right to withdraw)",
)
async def revoke_consent(
    payload: ConsentRevokeRequest,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """
    Revokes a consent type for a patient.

    Revocation is recorded as a new ConsentLog row (granted=False) — the original
    grant record is never deleted, preserving the full audit trail.

    Revoking `data_processing` blocks future triage sessions for this patient_id.
    Existing session records are retained (legal obligation) but no new processing
    is performed.

    Per DPDP §6(4), revocation is made as easy as granting — no authentication
    barrier beyond patient_id ownership.
    """
    _validate_consent_types([payload.consent_type])
    await _get_patient_or_404(payload.patient_id, db)

    ip_hash = _hash_ip(request.client.host if request.client else None)

    record = ConsentLog(
        patient_id=payload.patient_id,
        consent_type=payload.consent_type,
        granted=False,
        ip_address=ip_hash,
        version=CURRENT_CONSENT_VERSION,
    )
    db.add(record)

    audit = AuditLog(
        event_type="consent.revoked",
        entity_type="consent_log",
        entity_id=str(payload.patient_id),
        actor_id=str(payload.patient_id),
        ip_address=ip_hash,
    )
    db.add(audit)

    await db.commit()

    logger.info(
        "vaidya.consent.revoked",
        patient=str(payload.patient_id)[:8],
        type=payload.consent_type,
    )

    return {
        "revoked": payload.consent_type,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "message": (
            f"Consent for '{payload.consent_type}' has been revoked. "
            "Your existing triage history is retained as required by law. "
            "No new data will be processed for this purpose."
        ),
    }


# ── GET /consent/status/{patient_id} ──────────────────────────────────────────

@router.get(
    "/status/{patient_id}",
    response_model=ConsentStatusResponse,
    summary="Current effective consent state for a patient",
)
async def get_consent_status(
    patient_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """
    Returns the current effective consent for each purpose.

    'Effective' means the most recent record for each consent_type — if the
    latest record is granted=False, consent is considered revoked regardless
    of any earlier grants.

    The `has_required_consent` flag indicates whether `data_processing` is
    currently granted — the triage endpoint checks this before processing.
    """
    await _get_patient_or_404(patient_id, db)

    result = await db.execute(
        select(ConsentLog)
        .where(ConsentLog.patient_id == patient_id)
        .order_by(ConsentLog.timestamp.desc())
    )
    all_records = result.scalars().all()

    # Collapse to latest record per type
    seen: dict = {}
    for rec in all_records:
        if rec.consent_type not in seen:
            seen[rec.consent_type] = rec

    consents = [
        ConsentStatusItem(
            consent_type=ct,
            granted=rec.granted,
            version=rec.version,
            timestamp=rec.timestamp,
        )
        for ct, rec in seen.items()
    ]

    has_required = any(
        c.consent_type == "data_processing" and c.granted for c in consents
    )

    return ConsentStatusResponse(
        patient_id=patient_id,
        consents=consents,
        has_required_consent=has_required,
        evaluated_at=datetime.now(timezone.utc),
    )


# ── GET /consent/audit/{patient_id} ───────────────────────────────────────────

@router.get(
    "/audit/{patient_id}",
    response_model=ConsentAuditResponse,
    summary="Full immutable consent audit trail for CDSCO / DPDP compliance review",
)
async def get_consent_audit(
    patient_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user_id),   # auth required for audit access
):
    """
    Returns the full, immutable consent history for a patient in chronological order.

    Used by the DPO during compliance audits and by CDSCO inspectors to verify
    that informed consent was obtained before each processing activity.

    IP addresses are returned as a 16-char hash prefix — sufficient to detect
    anomalies (e.g. consent granted from a different device) without exposing PII.
    """
    await _get_patient_or_404(patient_id, db)

    result = await db.execute(
        select(ConsentLog)
        .where(ConsentLog.patient_id == patient_id)
        .order_by(ConsentLog.timestamp.asc())
    )
    records = result.scalars().all()

    entries = [
        ConsentAuditEntry(
            id=rec.id,
            consent_type=rec.consent_type,
            granted=rec.granted,
            version=rec.version,
            timestamp=rec.timestamp,
            ip_hash=rec.ip_address,  # already hashed at write time
        )
        for rec in records
    ]

    return ConsentAuditResponse(
        patient_id=patient_id,
        entries=entries,
        total=len(entries),
    )


# ── GET /consent/descriptions ─────────────────────────────────────────────────

@router.get(
    "/descriptions",
    summary="Return consent purpose descriptions in all supported languages",
)
async def get_consent_descriptions(
    language: str = "en",
):
    """
    Returns the human-readable description of each consent purpose.
    The frontend must display these to the patient before requesting consent.

    Descriptions are version-pinned — the `policy_version` in the response
    must be passed back in the grant request to prove the patient saw this version.
    """
    # Translations would be loaded from i18n files in production
    # For now, English is canonical; hi/ta stubs included for completeness
    return {
        "policy_version": CURRENT_CONSENT_VERSION,
        "language": language,
        "purposes": [
            {
                "type": ct,
                "required": ct == "data_processing",
                "description": desc,
            }
            for ct, desc in CONSENT_DESCRIPTIONS.items()
        ],
    }

File: frontend/src/services/consent.ts
/**
 * Vaidya — Consent service (descriptions, grant, revoke, status, audit)
 * Falls back to demo mode when the backend is unreachable.
 */

import { apiClient } from './api';

// ── Types ─────────────────────────────────────────────────────────────────────

export interface ConsentPurpose {
  type:        string;
  required:    boolean;
  description: string;
}

export interface ConsentDescriptionsResponse {
  policy_version: string;
  language:       string;
  purposes:       ConsentPurpose[];
}

export interface ConsentGrantBody {
  patient_id:      string;
  consent_types:   string[];
  policy_version?: string;
  language?:       string;
}

export interface ConsentGrantResponse {
  granted:        string[];
  policy_version: string;
  timestamp:      string;
  message:        string;
}

export interface ConsentRevokeBody {
  patient_id:   string;
  consent_type: string;
  reason?:      string;
}

export interface ConsentRevokeResponse {
  revoked:   string;
  timestamp: string;
  message:   string;
}

export interface ConsentRecord {
  consent_type: string;
  granted:      boolean;
  version:      string;
  timestamp:    string;
}

export interface ConsentStatusResponse {
  patient_id:           string;
  consents:             ConsentRecord[];
  has_required_consent: boolean;
  evaluated_at:         string;
}

export interface ConsentAuditEntry {
  id:           string;
  consent_type: string;
  granted:      boolean;
  version:      string;
  timestamp:    string;
  ip_hash:      string;
}

export interface ConsentAuditResponse {
  patient_id: string;
  entries:    ConsentAuditEntry[];
  total:      number;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

async function demoDelay(ms = 600): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

// ── Demo data ─────────────────────────────────────────────────────────────────

const DEMO_POLICY_VERSION = '1.2.0';

const DEMO_DESCRIPTIONS: ConsentDescriptionsResponse = {
  policy_version: DEMO_POLICY_VERSION,
  language:       'en',
  purposes: [
    {
      type:        'data_processing',
      required:    true,
      description: 'We process your health information to provide symptom triage and medical guidance. This data is stored securely and never sold to third parties.',
    },
    {
      type:        'anonymised_analytics',
      required:    false,
      description: 'Anonymised, non-identifiable health data helps detect disease outbreaks early and improve public health response in your district.',
    },
    {
      type:        'asha_contact',
      required:    false,
      description: 'Allow your local ASHA (Accredited Social Health Activist) worker to follow up with you for urgent or serious health concerns.',
    },
  ],
};

const DEMO_CONSENTS: ConsentRecord[] = [
  { consent_type: 'data_processing',      granted: true,  version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString() },
  { consent_type: 'anonymised_analytics', granted: true,  version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString() },
  { consent_type: 'asha_contact',         granted: false, version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString() },
];

const DEMO_AUDIT_ENTRIES: ConsentAuditEntry[] = [
  { id: 'audit-001', consent_type: 'data_processing',      granted: true,  version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString(), ip_hash: 'a3f8c2d1' },
  { id: 'audit-002', consent_type: 'anonymised_analytics', granted: true,  version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString(), ip_hash: 'a3f8c2d1' },
  { id: 'audit-003', consent_type: 'asha_contact',         granted: false, version: DEMO_POLICY_VERSION, timestamp: new Date(Date.now() - 30 * 86400 * 1000).toISOString(), ip_hash: 'a3f8c2d1' },
];

// ── Consent API ───────────────────────────────────────────────────────────────

export async function getConsentDescriptions(params?: {
  language?: string;
}): Promise<ConsentDescriptionsResponse> {
  try {
    const { data } = await apiClient.get('/consent/descriptions', { params, timeout: 8_000 });
    return data;
  } catch {
    await demoDelay();
    return { ...DEMO_DESCRIPTIONS, language: params?.language ?? 'en' };
  }
}

export async function grantConsent(body: ConsentGrantBody): Promise<ConsentGrantResponse> {
  try {
    const { data } = await apiClient.post('/consent/grant', body, { timeout: 8_000 });
    return data;
  } catch {
    await demoDelay(700);
    return {
      granted:        body.consent_types,
      policy_version: body.policy_version ?? DEMO_POLICY_VERSION,
      timestamp:      new Date().toISOString(),
      message:        'Consent recorded successfully.',
    };
  }
}

export async function revokeConsent(body: ConsentRevokeBody): Promise<ConsentRevokeResponse> {
  try {
    const { data } = await apiClient.post('/consent/revoke', body, { timeout: 8_000 });
    return data;
  } catch {
    await demoDelay(700);
    return {
      revoked:   body.consent_type,
      timestamp: new Date().toISOString(),
      message:   'Consent revoked. Your data preferences have been updated.',
    };
  }
}

export async function getConsentStatus(patient_id: string): Promise<ConsentStatusResponse> {
  try {
    const { data } = await apiClient.get(`/consent/status/${patient_id}`, { timeout: 8_000 });
    return data;
  } catch {
    await demoDelay(600);
    return {
      patient_id,
      consents:             DEMO_CONSENTS,
      has_required_consent: true,
      evaluated_at:         new Date().toISOString(),
    };
  }
}

export async function getConsentAudit(patient_id: string): Promise<ConsentAuditResponse> {
  try {
    const { data } = await apiClient.get(`/consent/audit/${patient_id}`, { timeout: 8_000 });
    return data;
  } catch {
    await demoDelay(700);
    return {
      patient_id,
      entries: DEMO_AUDIT_ENTRIES,
      total:   DEMO_AUDIT_ENTRIES.length,
    };
  }
}
